emma-flink/src/main/scala/org/emmalanguage/api/FlinkDataSet.scala

```diff
@@ -91,6 +97,47 @@ class FlinkDataSet[A: Meta] private[api]
     rep.distinct
 
   // -----------------------------------------------------
+  // Partition-based Ops
+  // -----------------------------------------------------
+
+  def sample(k: Int, seed: Long = 5394826801L): Vector[A] = {
+    // sample per partition and sorted by partition ID
+    val Seq(hd, tl@_*) = new DataSetUtils(rep).zipWithIndex
+      .mapPartition(new RichMapPartitionFunction[(Long, A), (Int, Array[Option[A]])] {
+        @transient private var pid = 0
+
+        override def open(parameters: Configuration) = {
+          super.open(parameters)
+          pid = getRuntimeContext.getIndexOfThisSubtask
+        }
+
+        import scala.collection.JavaConversions._
+
+        def mapPartition(it: lang.Iterable[(Long, A)], out: Collector[(Int, Array[Option[A]])]) = {
+          val sample = Array.fill(k)(Option.empty[A])
+          for ((i, e) <- it) {
+            if (i >= k) {
+              val j = util.RanHash(seed).at(i).nextLong(i + 1)
+              if (j < k) sample(j.toInt) = Some(e)
+            } else sample(i.toInt) = Some(e)
```

Comment:
I had to adapt the [original reservoir sampling procedure](https://en.wikipedia.org/wiki/Reservoir_sampling#Algorithm_R) to accommodate for the partitioned input.

In the case where we have one partition, we pick the first K elements and then replace a random element in the reservoir with the `i`-th element with a decreasing probability. As a reference, check the `ScalaSeq` implementation.

However, in the case of Flink / Spark, this sequential step done once per partition, and for `n` partitions we have to merge `n` `sample` lists. To determine a position `j` has been updated in partition `p`, I changed the container type from `A` to `Option[A]`. Merging then gives precedence to options from the rhs (which come are later).

Let me illustrate this with an example:


A `DataBag`

```scala
val xs = DataBag(Seq['a, 'b, 'c, 'd, 'e, 'f, 'g, 'h))
``` 
is distributed (in a non-deterministic way) across two partitions.

```scala
P(0) = Seq('c, 'd, 'a, 'h)
P(1) = Seq('b, 'e, 'g, 'f)
```

Running 

```scala
xs.sample(5)
``` 

will produce the following partial sample.

```scala
samples(0) = Vector(Some('c), Some('d), Some('a), Some('h), None)
```

The result here is deterministic, as the indices generated by `zipWithIndex` on partition `P(0)` are all `<= k` (they are in the `0 to 3` range). 

On `P(1)`, the first element in the sequence (`'b`) will be set on the fifth position of the `sample(1)` vector, and depending on the concrete sequence of random numbers, some of the positions may be overridden as well, e.g. we might end up with something like this.

```scala
samples(2) = Vector(None, Some('g), None, None, Some('b))
```

The final sample then is 

```scala
sample = samples.fold(sample(i)(_ ⨁ _)
```

where `xs ⨁ ys` merged two vectors element-wise, preferring `Some(y)` over `Some(x)`. In the concrete example, the final result looks as follows.

```scala
sample = Vector(Some('c), Some('g), Some('a), Some('h), Some('b))
```